src/foam/dao/pg/PostgresDAO.java

```diff
@@ -0,0 +1,368 @@
+package foam.dao.pg;
+
+import foam.core.ClassInfo;
+import foam.core.FObject;
+import foam.core.PropertyInfo;
+import foam.core.X;
+import foam.dao.AbstractDAO;
+import foam.dao.ListSink;
+import foam.dao.SQLType;
+import foam.dao.Sink;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+
+import java.sql.*;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class PostgresDAO
+    extends AbstractDAO
+{
+  protected ConnectionPool connectionPool = new ConnectionPool();
+  protected ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder b = super.get();
+      b.setLength(0);
+      return b;
+    }
+  };
+
+  protected final String table;
+  protected final List<PropertyInfo> props;
+
+  public PostgresDAO(ClassInfo of, String host, String port, String dbName, String username, String password) throws SQLException {
+    setOf(of);
+
+    if ( dbName == null || username == null ) {
+      throw new IllegalArgumentException("Illegal arguments");
+    }
+
+    host = (host != null) ? host : "localhost";
+    port = (port != null) ? port : "5432";
+
+    // set up connection pool
+    connectionPool.setup(host, port, dbName, username, password);
+
+    // load columns and sql types
+    table = of.getObjClass().getSimpleName().toLowerCase();
+    if ( ! createTable(of) ) {
+      throw new SQLException("Error creating table");
+    }
+
+    props = ((List<PropertyInfo>) of.getAxiomsByClass(PropertyInfo.class))
+        .stream().filter(e -> ! e.getStorageTransient())
+        .collect(Collectors.toList());
+  }
+
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    Connection c = null;
+    PreparedStatement stmt = null;
+    ResultSet resultSet = null;
+
+    try {
+      c = connectionPool.getConnection();
+
+      // select all if predicate is null, else use predicate
+      if ( predicate == null ) {
+        StringBuilder builder = sb.get()
+            .append("select * from ")
+            .append(table);
+        stmt = c.prepareStatement(builder.toString());
+      } else {
+        stmt = c.prepareStatement(predicate.createStatement(table));
+      }
+
+      resultSet = stmt.executeQuery();
+      while ( resultSet.next() ) {
+        sink.put(createFObject(resultSet), null);
+      }
+
+      return sink;
+    } catch (Exception e) {
```

Comment:
Catch Throwable.

Reply:
Modified all catch blocks to catch Throwables instead.